### ISSUE TITLE
Out of bounds `addr_of!` on invalid BTreeSet<Box<u8>>

### ISSUE THREAD

**Issue by 5225225** [rkyv/rkyv#307]
```rust
use std::collections::BTreeSet;

fn main() {
    let data = rkyv::util::AlignedBytes([1, 29, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 253, 0, 0, 116, 255, 255, 40, 0, 8, 0, 0, 0, 236, 255, 255, 255, 1, 128, 72, 0, 220, 255, 255, 255, 236, 255, 255, 255, 0, 0, 0, 0, 32, 0, 255, 254, 255, 0, 94, 2, 33, 0, 0, 0, 0, 0, 0, 0, 61, 1, 38, 0, 0, 32, 0, 255, 255, 1, 0, 1, 255, 255, 0, 184, 4, 0, 28, 0, 8, 0, 2, 142, 255, 255, 255, 3, 1, 255, 251, 0, 184, 255, 255, 255]);

    rkyv::from_bytes::<BTreeSet<Box<u8>>>(&*data);
}
```

I actually figured out what the issue with miri was. Miri had minimally aligned locals, which means I was running into the alignment check early, and nothing else happened.

This was found with address sanitizer, and when run with `MIRIFLAGS="-Zmiri-disable-stacked-borrows" cargo miri run` (because miri runs into stacked borrows errors earlier), gives

```rust
error: Undefined Behavior: dereferencing pointer failed: alloc877 has size 96, so pointer to 1024 bytes starting at offset 4 is out-of-bounds
   --> /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/collections/btree_map/validation.rs:440:24
    |
440 |         let tail_ptr = ptr::addr_of!((*value).tail) as *const LeafNodeEntry<K, V>;
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ dereferencing pointer failed: alloc877 has size 96, so pointer to 1024 bytes starting at offset 4 is out-of-bounds
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: backtrace:
    = note: inside `rkyv::collections::btree_map::validation::<impl bytecheck::CheckBytes<rkyv::validation::validators::DefaultValidator> for rkyv::collections::btree_map::Node<[rkyv::collections::btree_map::LeafNodeEntry<rkyv::boxed::ArchivedBox<u8>, ()>]>>::check_bytes` at /home/jess/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:1999:5
    = note: inside `rkyv::collections::btree_map::validation::<impl rkyv::collections::btree_map::NodeHeader>::manual_check_contents::<rkyv::boxed::ArchivedBox<u8>, (), rkyv::validation::validators::DefaultValidator>` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/collections/btree_map/validation.rs:346:13
    = note: inside `rkyv::collections::btree_map::validation::<impl rkyv::collections::btree_map::NodeHeader>::manual_check_bytes::<rkyv::boxed::ArchivedBox<u8>, (), rkyv::validation::validators::DefaultValidator>` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/collections/btree_map/validation.rs:288:9
    = note: inside `rkyv::collections::btree_map::validation::_::<impl bytecheck::CheckBytes<rkyv::validation::validators::DefaultValidator> for rkyv::collections::ArchivedBTreeMap<rkyv::boxed::ArchivedBox<u8>, ()>>::check_bytes` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/collections/btree_map/validation.rs:521:33
    = note: inside `rkyv::collections::btree_set::_::<impl bytecheck::CheckBytes<rkyv::validation::validators::DefaultValidator> for rkyv::collections::btree_set::ArchivedBTreeSet<rkyv::boxed::ArchivedBox<u8>>>::check_bytes` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/collections/btree_set.rs:11:32
    = note: inside `rkyv::validation::internal_check_archived_value_with_context::<std::collections::BTreeSet<std::boxed::Box<u8>>, rkyv::validation::validators::DefaultValidator>` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/validation/mod.rs:324:13
    = note: inside `rkyv::check_archived_root_with_context::<std::collections::BTreeSet<std::boxed::Box<u8>>, rkyv::validation::validators::DefaultValidator>` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/validation/mod.rs:364:5
    = note: inside `rkyv::check_archived_root::<std::collections::BTreeSet<std::boxed::Box<u8>>>` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/validation/validators/mod.rs:228:5
    = note: inside `rkyv::from_bytes::<std::collections::BTreeSet<std::boxed::Box<u8>>>` at /home/jess/.cargo/registry/src/github.com-1ecc6299db9ec823/rkyv-0.7.39/src/validation/validators/util.rs:70:5
note: inside `main` at src/main.rs:6:5
   --> src/main.rs:6:5
    |
6   |     rkyv::from_bytes::<BTreeSet<Box<u8>>>(&*data);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `ptr::addr_of` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by 5225225** [comment]
This is actually caught by the existing invalid BTree test. I'll work on adding miri support (currently, if you run miri tests, the crate fails quite quickly), it would be good to get at least a basic miri run as part of CI here.

**Comment by djkoloski** [comment]
I usually run MIRI in a couple configurations before releasing a new version, the details are in ` release_checklist.md`. rkyv actually used to pass MIRI, but doesn't any more due to provenance narrowing when making a reference. This has made it more difficult to run, but there's probably a configuration that turns off this provenance narrowing behavior. A run in CI would also be helpful.

**Comment by djkoloski** [comment]
Fixed by 64b9f4ddb29a535428fddf1c760c4c98ad719819
